### PR TITLE
feat: Add BaseWriter, and example NIFTIWriter with notebooks documenting them

### PR DESCRIPTION
**Pull request by jjjermiah**
Mostly inspired and reusing logic from Med-ImageTools `DICOMSorter` design

- I will very likely look to extract the `BaseWriter` stuff into MIT later so it can be reused around, but worth going through a review process 

## Summary by CodeRabbit

- **New Features**
  - Enhanced Jupyter notebooks for saving medical imaging data in NIFTI format and CSV metadata.
  - Added examples for using subclasses of `BaseWriter` for writing text and CSV files.
  - Introduced the `NIFTIWriter` class for managing NIFTI file writing with validation and error handling.
  - Added new `CSVWriter` class for saving data in CSV format.

- **Documentation**
  - Added comprehensive documentation for the Writer system, including details on `BaseWriter` and its subclasses.
  - Updated module docstrings for better clarity on their purpose.

- **Bug Fixes**
  - Improved error handling for file writing operations.

- **Chores**
  - Adjusted linting configuration to include and exclude specific files and directories.

### CODE REVIEW

File: src/readii/io/writers/base_writer.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+from readii.utils import logger
+
+
+@dataclass
+class PatternResolver:
+	"""Handles parsing and validating filename patterns."""
+
+	filename_format: str = field(init=True)
+
+	DEFAULT_PATTERN: ClassVar[re.Pattern] = re.compile(r"%(\w+)|\{(\w+)\}")
```
Comment (strixy16) [inline src/readii/io/writers/base_writer.py:21]: Can you add a comment breaking down what this regex is at a high level?

File: src/readii/io/writers/base_writer.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		opening connections or preparing resources.
+		"""
+		logger.debug(f"Entering context manager for {self.__class__.__name__}")
+		return self
+
+	def __exit__(
+		self: "BaseWriter",
+		exc_type: Optional[type],
+		exc_value: Optional[BaseException],
+		traceback: Optional[TracebackType],
```
Comment (strixy16) [inline src/readii/io/writers/base_writer.py:144]: `traceback` isn't used, is it something that might be used in a subclass?

File: src/readii/io/writers/nifti_writer.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		try:
+			sitk.WriteImage(
+				image, str(out_path), useCompression=True, compressionLevel=self.compression_level
+			)
+		except Exception as e:
+			msg = f"Error writing image to file {out_path}: {e}"
+			raise NiftiWriterIOError(msg) from e
+		else:
+			logger.info("Image saved successfully.", out_path=out_path)
+			return out_path
```
Comment (strixy16) [inline src/readii/io/writers/nifti_writer.py:118]: This looks good, but if I want to save a NIFTI file from a numpy array, do I need to make another NIFTIWriter class, another save function, or would it need to be handled in this single save function?

File: src/readii/io/writers/nifti_writer.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+	nifti_writer = NIFTIWriter(
+		root_directory=Path("TRASH", "nifti_writer_examples"),
+		filename_format="{NegativeControl}_{Region}/{SubjectID}_{Modality}.nii.gz",
+		compression_level=9,
+		overwrite=False,
+		create_dirs=True,
+	)
+
+	print(nifti_writer)
```
Comment (strixy16) [inline src/readii/io/writers/nifti_writer.py:132]: Is this safe to have here?

**Review by strixy16 — CHANGES_REQUESTED** [review]
Looks good! 

Only concern is in how a BaseWriter subclass save function handles different types of input files to the save function - see comment in NIFTIWriter save function.

File: src/readii/io/writers/nifti_writer.py (hunk 3 above)
Comment (jjjermiah) [reply src/readii/io/writers/nifti_writer.py:118]: that is a great point. do you know if its just converting using `ImageFromArray` or is there more steps required? if its simple enough then I can refactor this to take in either and handle it before saving

File: src/readii/io/writers/nifti_writer.py (hunk 4 above)
Comment (jjjermiah) [reply src/readii/io/writers/nifti_writer.py:132]: safe as in wont affect code at runtime yes, i left it here for future 'documentation' purposes, will port over to the docs in the future

File: src/readii/io/writers/base_writer.py (hunk 1 above)
Comment (jjjermiah) [reply src/readii/io/writers/base_writer.py]: yup, moved to a utils directory and added more documentation throughout the module

Let me know if this is not confusing:

https://github.com/bhklab/readii/blob/cdfcdfcf1ef54cf3aeaa20d94cfaf3610f49792d/src/readii/io/utils/pattern_resolver.py

File: src/readii/io/utils/pattern_resolver.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import re
+from dataclasses import dataclass, field
+from typing import Any, ClassVar, Dict, Tuple
+
+from imgtools.dicom.sort.exceptions import InvalidPatternError  # type: ignore
+from imgtools.dicom.sort.parser import PatternParser  # type: ignore
```
[the remark this answers is not in the document]
Comment (jjjermiah) [reply src/readii/io/utils/pattern_resolver.py:6]: no this package is meant to work off of med-imagetools

File: src/readii/io/writers/nifti_writer.py (hunk 3 above)
Comment (strixy16) [reply src/readii/io/writers/nifti_writer.py:114]: We can just do that yeah.

File: src/readii/io/writers/nifti_writer.py (hunk 4 above)
Comment (strixy16) [reply src/readii/io/writers/nifti_writer.py:128]: Gotcha.

**Review by strixy16 — APPROVED** [review]

**Review by strixy16 — APPROVED** [review]

**Pull request merged**